src/shogun/optimization/StanFirstOrderSAGCostFunction.h

```diff
@@ -0,0 +1,177 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Elfarouk
+ */
+
+#ifndef StanFirstOrderSAGCostFunction_H
+#define StanFirstOrderSAGCostFunction_H
+
+#include <stan/math.hpp>
+#include <functional>
+#include <shogun/lib/SGMatrix.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/base/range.h>
+#include <shogun/lib/config.h>
+#include <shogun/mathematics/eigen3.h>
+#include <shogun/optimization/FirstOrderSAGCostFunction.h>
+#include <shogun/mathematics/linalg/LinalgNamespace.h>
+#include <shogun/mathematics/linalg/LinalgBackendViennaCL.h>
+
+using StanVector = Eigen::Matrix<stan::math::var, Eigen::Dynamic, 1>;
+template <class T>
+using FunctionReturnsStan = std::function<stan::math::var(T)>;
+template <class R>
+using FunctionReturnsStanVectorArg = std::function<stan::math::var(const StanVector&, R)>;
+template <class S>
+using StanFunctionsVector =
+    Eigen::Matrix<FunctionReturnsStanVectorArg<S>, Eigen::Dynamic, 1>;
+namespace shogun
+{
+	/** @brief The first order stochastic cost function base class for
+	 * implementing the SAG Cost function
+	 *
+	 * The class gives the implementation used in first order stochastic
+	 * minimizers
+	 *
+	 * The cost function must be Written as a finite sample-specific sum of
+	 * cost.
+	 * For example, least squares cost function,
+	 * \f[
+	 * f(w)=\frac{ \sum_i{ (y_i-w^T x_i)^2 } }{2}
+	 * \f]
+	 * where \f$(y_i,x_i)\f$ is the i-th sample,
+	 * \f$y_i\f$ is the label and \f$x_i\f$ is the features
+	 */
+	class StanFirstOrderSAGCostFunction : public FirstOrderSAGCostFunction
```

Comment:
Not sure if this relationship between `StanFirstOrderSAGCostFunction` and `FirstOrderSAGCostFunction` makes sense.

Reply:
Agreed. I've changed the parent class to be FirstOrderStochasticCostFunction, since all these implemented functions are directly from it. (FirstOrderSAGCostFunction also inherits from FirstOrderStochasticCostFunction so it makes sense here to inherit from it as the Stan version is an alternative of FirstOrderSAGCostFunction).